pkg/controller/kube_controller.go

```diff
@@ -0,0 +1,138 @@
+
+	"github.com/hashicorp/go-retryablehttp"
+	"github.com/jetstack/version-checker/pkg/metrics"
+	"github.com/jetstack/version-checker/pkg/version/semver"
+)
+
+const channelURLSuffix = "https://dl.k8s.io/release/"
+
+type ClusterVersionScheduler struct {
+	client   kubernetes.Interface
+	log      *logrus.Entry
+	metrics  *metrics.Metrics
+	interval time.Duration
+	channel  string
+}
+
+func NewKubeReconciler(
+	log *logrus.Entry,
+	config *rest.Config,
+	metrics *metrics.Metrics,
+	interval time.Duration,
+	channel string,
+) *ClusterVersionScheduler {
+
+	return &ClusterVersionScheduler{
+		log:      log,
+		client:   kubernetes.NewForConfigOrDie(config),
+		interval: interval,
+		metrics:  metrics,
+		channel:  channel,
+	}
+}
+
+func (s *ClusterVersionScheduler) Start(ctx context.Context) error {
+	go s.runScheduler(ctx)
+	return s.reconcile(ctx)
+}
+
+func (s *ClusterVersionScheduler) runScheduler(ctx context.Context) {
+	ticker := time.NewTicker(s.interval)
+	defer ticker.Stop()
+
+	s.log.WithField("interval", s.interval).WithField("channel", s.channel).
+		Info("ClusterVersionScheduler started")
+
+	for {
+		select {
+		case <-ctx.Done():
+			s.log.Info("ClusterVersionScheduler stopping")
+			return
+		case <-ticker.C:
+			if err := s.reconcile(ctx); err != nil {
+				s.log.Error(err, "Failed to reconcile cluster version")
+			}
+		}
+	}
+}
+
+func (s *ClusterVersionScheduler) reconcile(_ context.Context) error {
+	// Get current cluster version
+	current, err := s.client.Discovery().ServerVersion()
+	if err != nil {
+		return fmt.Errorf("getting cluster version: %w", err)
+	}
+
+	// Get latest stable version
+	latest, err := getLatestStableVersion(s.channel)
+	if err != nil {
+		return fmt.Errorf("fetching latest stable version: %w", err)
+	}
+
+	latestSemVer := semver.Parse(latest)
+	currentSemVer := semver.Parse(current.GitVersion)
+
+	// Create version strings without metadata for comparison
+	currentSemVerNoMeta := fmt.Sprintf("%d.%d.%d", currentSemVer.Major(), currentSemVer.Minor(), currentSemVer.Patch())
+	latestSemVerNoMeta := fmt.Sprintf("%d.%d.%d", latestSemVer.Major(), latestSemVer.Minor(), latestSemVer.Patch())
+
+	// Parse the versions without metadata for comparison
+	currentComparable := semver.Parse(currentSemVerNoMeta)
+	latestComparable := semver.Parse(latestSemVerNoMeta)
+
+	// Register metrics!
+	s.metrics.RegisterKubeVersion(!currentComparable.LessThan(latestComparable),
+		currentSemVerNoMeta, latestSemVerNoMeta,
+		s.channel,
+	)
+
+	s.log.WithFields(logrus.Fields{
+		"currentVersion": currentSemVerNoMeta,
+		"latestStable":   latestSemVerNoMeta,
+		"channel":        s.channel,
+	}).Info("Cluster version check complete")
+
+	return nil
+}
+
+func getLatestStableVersion(channel string) (string, error) {
+	if !strings.HasSuffix(channel, ".txt") {
+		channel += ".txt"
+	}
+
+	// We don't need a `/` here as its should be in the channelURLSuffix
+	channelURL := fmt.Sprintf("%s%s", channelURLSuffix, channel)
```

Comment:
Nit: It may be better to use the URI or path joining over `fmt.Sprintf` https://pkg.go.dev/net/url#JoinPath perhaps?